sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/highlighting/XMLHighlighting.java

```diff
@@ -0,0 +1,213 @@
+    InputStream inputStream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
+    return getHighlightingData(inputStream, new SourceFileOffsets(content));
+  }
+
+  public static List<HighlightingData> getHighlightingData(InputStream inputStream, SourceFileOffsets sourceFileOffsets) throws XMLStreamException {
+    XMLStreamReader xmlReader = XMLInputFactory.newInstance().createXMLStreamReader(inputStream);
+    List<HighlightingData> highlighting = new ArrayList<>();
+    highlightXmlDeclaration(sourceFileOffsets, highlighting);
+
+    while (xmlReader.hasNext()) {
+      Location prevLocation = xmlReader.getLocation();
+      xmlReader.next();
+      int startOffset = sourceFileOffsets.startOffset(xmlReader.getLocation().getLineNumber(), xmlReader.getLocation().getColumnNumber());
+
+      switch (xmlReader.getEventType()) {
+        case XMLStreamConstants.START_ELEMENT:
+          highlightStartElement(sourceFileOffsets, xmlReader, highlighting, startOffset);
+          break;
+
+        case XMLStreamConstants.END_ELEMENT:
+          highlightEndElement(sourceFileOffsets, xmlReader, highlighting, prevLocation, startOffset);
+          break;
+
+        case XMLStreamConstants.CDATA:
+          highlightCData(sourceFileOffsets, highlighting, startOffset);
+          break;
+
+        case XMLStreamConstants.DTD:
+          highlightDTD(sourceFileOffsets, highlighting, startOffset);
+          break;
+
+        case XMLStreamConstants.COMMENT:
+          // 7 is length of opening and closing symbols : "<!--" and "-->"
+          highlighting.add(new HighlightingData(startOffset, startOffset + xmlReader.getTextLength() + 7, "j"));
+          break;
+
+        default:
+          break;
+      }
+    }
+
+    return highlighting;
+  }
+
+  private static void highlightDTD(SourceFileOffsets sourceFileOffsets, List<HighlightingData> highlighting, int startOffset) {
+    int closingBracketStartOffset;
+    closingBracketStartOffset = getTagClosingBracketStartOffset(startOffset, sourceFileOffsets.content());
+    highlighting.add(new HighlightingData(startOffset, startOffset + 9, "j"));
+    highlighting.add(new HighlightingData(closingBracketStartOffset, closingBracketStartOffset + 1, "j"));
+  }
+
+  private static void highlightCData(SourceFileOffsets sourceFileOffsets, List<HighlightingData> highlighting, int startOffset) {
+    int closingBracketStartOffset = getCDATAClosingBracketStartOffset(startOffset, sourceFileOffsets.content());
+
+    // 9 is length of "<![CDATA["
+    highlighting.add(new HighlightingData(startOffset, startOffset + 9, "k"));
+
+    // highlight "]]>"
+    highlighting.add(new HighlightingData(closingBracketStartOffset - 2, closingBracketStartOffset + 1, "k"));
+  }
+
+  private static void highlightEndElement(SourceFileOffsets sourceFileOffsets, XMLStreamReader xmlReader, List<HighlightingData> highlighting, Location prevLocation, int startOffset) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Split this 184 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)